### PR TITLE
enhancement(buffers): add support for codec/schema evolution

### LINKED ISSUE #10939: enhancement(buffers): update disk v2 record format to include version field to support schema evolution
As part of designing the new version of disk buffers (commonly referred to as `disk v2`), we want to make sure that the on-disk format has some ability to be extended over time in a backwards/forwards-compatible way.

A straightforward way to achieve this would be to include a version field in the record archival type, to describe how to decode the inner record.  Everything else could stay as is -- length delimiter, the archival record type having its checksum, ID, etc -- but the version field could simply function much like an integer map of "version => codec".

The intention for this would be to eventually move away from Protocol Buffers to something `rkyv`-based -- once `rkyv` has a suitable path towards schema evolution -- where version 1 (currrent) would be Protocol Buffers, and version 2 would eventually be `rkyv`-based.

The reader would simply skip versions that did not much the version expected for the given Vector binary, so we would still have to provide deprecation schedules around removing support for older record versions so that people could upgrade, process older messages, upgrade again, and then be on the latest record version, etc.

### PR DESCRIPTION
**Pull request by tobz**
This PR forms the basis of adding in the necessary components for allowing versioning and schema evolution of buffered events in future versions of Vector.

Closes #10939.

## Context

Currently, all events that flow through a buffer are either kept in their native Rust representation when using in-memory buffers, or are serialized to Protocol Buffers when using disk buffers.  While events cannot possibly change their Rust representation at runtime, events could change over time between different Vector versions which is why Protocol Buffers are employed for Vector-to-Vector communication as well as serialization before being stored in disk buffers.  Protocol Buffers already provide a mechanism for schema evolution which makes it attractive for this purpose.

However, as we continue to push harder and harder on the performance envelope of Vector, we may need to consider alternative serialization approaches.  This represents a constraint as we want to, wherever possible, maintain compatibility between Vector versions so that operators are not forced to follow complicated upgrade processes, or otherwise spend an outsized amount of time ensuring that they don't lose data.  A part of this is ensuring that we have support the versioning of records stored in disk buffers so that we can provide schema evolution that is decoupled with the codec itself, as there is evidence that Protocol Buffers encoding/decoding performance is currently a limiting factor on the performance of our new disk buffer implementation.

## What is versioning? What is schema evolution?

Support for versioning, or schema evolution, is the mechanism by which one piece of code can understand and handle multiple "versions" of the same logical piece of data.  For example, one version of Vector maybe encode an event and end up writing 10 fields, and the next version of Vector may instead write 9 fields, or 11 fields, or some other amount.  Future versions of Vector should, wherever possible, be able to understand the payload whether it has 10 fields or 9 fields or however many fields.  This is schema evolution.

Even further, we may want to change codecs entirely, rather than simply the shape of an object itself, which is why we need versioning at the buffer level, rather than as a fundamental part of the codec of choice.

## Solution

This PR adds support for versioning/schema evolution by adding the infrastructure to provide codec "metadata" at runtime, that can be bundled with an encoded record and then used during decoding to understand how a record should be decoded.

We've removed the separate `EncodeBytes` and `DecodeBytes` traits -- as we never really had a need for asymmetrical encoding/decoding -- and have replaced them with a single trait -- `Encodable` -- which encapsulates the idea of providing codec metadata and encoding/decoding all in one.

The process now looks like this:
- take an item that implements `Encodable` and use it to do the encoding (same)
- grab the metadata from that item and store with with the encoding item (new)
- read both the metadata and the encoded item when you wish to decode (semi-new)
- ask the decoder if they support an item encoded with the given metadata (new)
- if it does, pass the encoded item and the metadata together for decoding (semi-new)

Essentially, any implementation of `Encodable` will always provide metadata that describes how an item will be encoded _right now_, in this particular Vector process.  When that record, or any other record encoded under the new `Encodable` regimen, is read off of disk, the caller can first check if that record is still decodable by passing the metadata to `Encodable::can_decode`.  This allows an implementation of `Encodable` to support however many versions of a codec/schema are needed at any one moment, allowing for graceful deprecation over time, etc.  If the record is deemed decodable, decoding happens normally after that, with the difference being that the metadata is also passed, so the decoder knows exactly which codec, schema, etc, must be used for _this_ particular record.

## Implementation Details

Metadata is modeled as an associated type on `Encodable` which must satisfy a new trait: `AsMetadata`.  This trait forces the user to be able to convert the metadata to a `u32`, and to fallibly attempt converting a `u32` to the metadata type.

This allows us to store a small, fixed-size value with each record while still providing a large amount of space for defining the metadata itself.  Ostensibly, metadata will consist of flags that define certain characteristics about the record: which codec we used, which version of the event schema is being used, and so on.  In practice, using the `enumflags2` crate which lets us define true bitflags, this gives us 32 distinct flags to define and combine.  While we don't anticipate needing to ever use them all, a `u32` was a convenient choice as it will fit within the existing alignment/padding that we use for serialized records on disk in the disk buffer v2 implementation.

If an implementor wanted to, the `u32` could also be split into multiple fields: any combination of `u8` and `u16`, to use as a more traditional monotonic version identifier, and so on.

## Caveats

As we've replaced `EncodeBytes`/`DecodeBytes`, this naturally affects all code which used those traits, and that includes the disk buffer v1 implementation.  Currently, the code has been updated such that we still default to only using Protocol Buffers, which makes this a no-op in encoding/decoding behavior for the disk buffer v1 implementation.  We've hardcoded the metadata for `Event`, and have added unit tests and loud, angry sounding code comments, that explicitly state that the metadata cannot be changed until the disk buffer v1 implementation is entirely gone.  While we may still support decoding Protocol Buffers even after disk buffer v1 is gone, we are committing to not changing the default codec/schema/etc until that happens, as a means to preserve disk buffer v1 compatibility between releases.

## Testing Plan

New tests / testing plan for this change:
- [x] unit test that asserts the metadata for `impl Encodable for Event` exactly matches `<flags that say we're only using protobuf>`
- [x] unit test that encodes `Event` via the raw `prost::Message` methods and then decodes via `Encodable`, and also in the opposite direction, to ensure our change from using only the raw `prost::Message` calls to `Encodable` are a no-op
- [x] write items to a disk buffer v1 on a version of Vector before this PR, and read them without issue using a version of Vector built off this PR

Signed-off-by: Toby Lawrence <toby@nuclearfurnace.com>

### CODE REVIEW

File: lib/vector-buffers/src/disk_v2/reader.rs [hunk 1]
```diff
@@ -305,7 +329,17 @@ where
         // - `try_next_record` does all the archive checks, checksum validation, etc
         let archived_record = unsafe { archived_root::<Record<'_>>(&self.aligned_buf) };
 
-        T::decode(archived_record.payload()).context(DecodeSnafu)
+        // Try and convert the raw record metadata into the true metadata type used by `T`, and then
+        // also verify that `T` is able to decode records with the metadata used for this record in particular.
+        let metadata =
+            T::Metadata::from_u32(archived_record.metadata()).ok_or(ReaderError::Incompatible)?;
```
Comment (spencergilbert) [inline lib/vector-buffers/src/disk_v2/reader.rs:335]: Would we ever want separate errors between this step and the `can_decode` step?

File: lib/vector-buffers/src/disk_v2/record.rs [hunk 2]
```diff
@@ -99,27 +111,36 @@ where
 
 impl<'a> Record<'a> {
     /// Creates a [`Record`] from the ID and payload, and calculates the checksum.
-    pub fn with_checksum(id: u64, payload: &'a [u8], checksummer: &Hasher) -> Self {
-        let checksum = generate_checksum(checksummer, id, payload);
+    pub fn with_checksum(id: u64, metadata: u32, payload: &'a [u8], checksummer: &Hasher) -> Self {
+        let checksum = generate_checksum(checksummer, id, metadata, payload);
```
Comment (spencergilbert) [inline lib/vector-buffers/src/disk_v2/record.rs:115]: Will updating the checksum here have any impact to existing v2 disk buffers?

**Review by spencergilbert — COMMENTED** [review]
Above my head for real approval but 2 questions

File: lib/vector-buffers/src/disk_v2/reader.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-buffers/src/disk_v2/reader.rs:335]: We might! I did have that thought, or at least the thought that there could bean additional `reason` field describing the "this metadata isn't even valid" vs "we don't support this record".  Maybe I'll just go ahead and implement that.

File: lib/vector-buffers/src/disk_v2/record.rs (hunk 2 above)
Comment (tobz) [reply lib/vector-buffers/src/disk_v2/record.rs:115]: Yes, but so does adding the metadata field itself.  This is definitely a breaking change, but we're intentionally doing it before even announcing the disk buffer v2 stuff as beta.

**Review by blt — APPROVED** [review]

File: lib/vector-buffers/src/disk_v2/reader.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-buffers/src/disk_v2/reader.rs]: Addressed this in 006a597.

**Review by jszwedko — COMMENTED** [review]
Nice! Happy to see this get in ahead of the next release. Thanks for including the new test cases.

**Pull request merged**